Python: Add updates, moves and deletes

Fokko opened the pull request.

rdblue commented on python/mkdocs/docs/api.md:
````diff
@@ -150,32 +150,25 @@ catalog.create_table(
 ```
 
-Or, use the UpdateSchema API directly:
+Use the Transaction API:
````
What about "Alternatively, use the transaction API to combine changes from multiple operations"? Otherwise it isn't clear why you'd use a transaction.

Fokko replied:
I tried to indicate that by also setting a property in the same transaction, but I agree that it is better to explicitly state it. Thanks!

Fokko replied:
```suggestion
Alternatively, use the transaction API to combine changes from multiple operations:
```

rdblue commented on python/pyiceberg/table/__init__.py:
```diff
@@ -225,17 +221,13 @@ def commit_transaction(self) -> Table:
         # Strip the catalog name
         if len(self._updates) > 0:
-            response = self._table.catalog._commit_table(  # pylint: disable=W0212
+            self._table._do_commit(  # pylint: disable=W0212
```
I'm fine with this being private for now, but we may want to actually expose it later.

rdblue commented on python/pyiceberg/table/__init__.py:
```diff
@@ -527,8 +519,13 @@ def history(self) -> List[SnapshotLogEntry]:
+    def update_schema(self, allow_incompatible_changes: bool = False, case_sensitive: bool = True) -> UpdateSchema:
+        return UpdateSchema(self, allow_incompatible_changes=allow_incompatible_changes, case_sensitive=case_sensitive)
+
+    def _do_commit(self, request: CommitTableRequest) -> None:
```
I think it would make more sense to pass requirements and updates. The identifier is already in the table and this can create the request. Or maybe the request should be created in `catalog._commit_table`?

Fokko replied:
Yes, I think that's cleaner as well, updated!

rdblue commented on python/pyiceberg/table/__init__.py:
```diff
@@ -934,206 +961,608 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
+        path = (path,) if isinstance(path, str) else path
+
+        if "." in path[-1]:
+            raise ValueError(f"Cannot add column with ambiguous name: {path[-1]}, provide a tuple instead")
```
This isn't quite correct. It's actually fine for the name to contain a `.` if the path was a tuple. If it was a tuple, then we know exactly what part is the parent. So this check is only needed if path is a str:

```python
if isinstance(path, str):
    if '.' in path:
        raise ValueError(...)
    path = (path,)

...
```

rdblue commented on python/pyiceberg/table/__init__.py:
```diff
@@ -934,206 +963,611 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
+        exists = False
+        try:
+            exists = self._schema.find_field(full_name, self._case_sensitive) is not None
+        except ValueError:
```
Why would `find_field` throw a `ValueError`? Shouldn't it return None instead?

Fokko replied:
```python
def find_field(self, name_or_id: Union[str, int], case_sensitive: bool = True) -> NestedField:
    """Find a field using a field name or field ID.

    Args:
        name_or_id (Union[str, int]): Either a field name or a field ID.
        case_sensitive (bool, optional): Whether to perform a case-sensitive lookup using a field name. Defaults to True.

    Raises:
        ValueError: When the value cannot be found.

    Returns:
        NestedField: The matched NestedField.
    """
```

Fokko replied:
It throws an error indeed.

Fokko replied:
This was in there before any PyIceberg release. This follows the _Easier to Ask Forgiveness Than Permission_ style, which is quite popular in Python. I personally like to avoid this pattern and avoid the exceptions. The more now we have the walrus `:=` operator.

rdblue replied:
Agreed. I guess it's established already though.

rdblue commented on python/pyiceberg/table/__init__.py:
```diff
@@ -934,206 +963,611 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
+        new_type = assign_fresh_schema_ids(field_type, self.assign_new_column_id)
+        field = NestedField(field_id=new_id, name=name, field_type=new_type, required=required, doc=doc)
+
+        self._adds[parent_id] = self._adds.get(parent_id, []) + [field]
```
Maybe it's me, but using + makes me wonder about the argument types. I find this more clear:

```python
self._adds[parent_id] = tuple(*self._adds.get(parent_id, []), field)
```

Fokko replied:
I'm open to making it more readable, but that doesn't work on my machine:
```python
➜  python git:(fd-followup) python3
Python 3.11.4 (main, Jul 25 2023, 17:36:13) [Clang 14.0.3 (clang-1403.0.22.14.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> tuple(*[1, 2], 3)
```

What does work is:
```python
self._adds[parent_id] = self._adds.get(parent_id, tuple()), + (field,)
```

But then we end up with the same thing as before, but then a tuple. I would suggest refactoring it as:

```python
if parent_id in self._adds:
    self._adds[parent_id].append(field)
else:
    self._adds[parent_id] = [field]
```

Then we don't recreate the collections all the time (but this isn't the hottest path, probably)

rdblue commented on python/pyiceberg/table/__init__.py:
```diff
@@ -934,206 +963,611 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
+            The UpdateSchema with the rename operation staged.
+        """
+        name_from = tuple(path_from.split(".")) if isinstance(path_from, str) else path_from
+        new_name = new_name.split(".")[-1]
```
In the Java API, this is the leaf field name. There's no need to split.

`rename("a.b", "c.d")` should produce field `a.c.d`

Fokko replied:
This is in Python as well, but we need to full path for the identifier columns 👍🏻

rdblue commented on python/pyiceberg/table/__init__.py:
```diff
@@ -934,206 +963,611 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
-        """Apply the pending changes to the original schema and returns the result.
+        if path_from in self._identifier_field_names:
+            self._identifier_field_names.remove(full_name_from)
+            self._identifier_field_names.add(f"{'.'.join(parent_path)}.{new_name}")
```
Ah, I see why you wanted to parse `name_from`. The problem is that we can never parse full names using `split` because we don't know whether `.` is a delimiter or a character in a name. In this case, you should look up the parent's name by finding the parent's ID and looking it up in the ID to name index.

By the way, I think this fixes a problem in the Java implementation. The `newName` is a leaf name, not a full name.

Fokko replied:
We had the same train of thought, I've updated this to:
```python
        if path_from in self._identifier_field_names:
            self._identifier_field_names.remove(path_from)
            self._identifier_field_names.add(f"{path_from[:-len(field_from.name)]}{new_name}")
```

rdblue commented on python/pyiceberg/table/__init__.py:
```diff
@@ -934,206 +963,611 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
+            The UpdateSchema with the requirement change staged.
         """
-        return _apply_changes(self._schema, self._adds, self._identifier_field_names)
+        self._set_column_requirement(path, True)
```
I don't think this invocation is clear. When passing a boolean arg, can you use a named argument?

Fokko replied:
My IDEA sets this:
<img width="592" alt="image" src="https://github.com/apache/iceberg/assets/1134248/8c0886df-3b87-436b-aa1c-2a2ca5ddc47c">

But I agree that `kwargs > args`, thanks for pointing this out

rdblue commented on python/pyiceberg/table/__init__.py:
```diff
@@ -934,206 +963,611 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
 
-            if exist_field:
-                raise ValueError(f"Cannot add column, name already exists: {name}")
+    def update_column(self, path: Union[str, Tuple[str, ...]], field_type: IcebergType) -> UpdateSchema:
```
Rather than having separate methods like the Java implementation, what about using a combined `update_column` with named parameters?

```python
def update_column(self, path: Union[str, Tuple[str, ...]], field_type: Optional[IcebergType] = None, required: Optional[boolean] = None, doc: Optional[str] = None) -> UpdateSchema:
     # find the column ...
     field_type = field_type if field_type is None else field.field_type
     doc = doc if doc is None else field.doc
     required = required if required is None else field.required
     ...
```

Fokko replied:
I like it, updated!

Fokko replied:
I left the old methods in place, WDYT?

rdblue commented on python/pyiceberg/table/__init__.py:
```diff
@@ -934,206 +963,611 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
+    def _find_for_move(self, name: str) -> Optional[int]:
+        try:
+            return self._schema.find_field(name, self._case_sensitive).field_id
+        except ValueError:
```
Flagging this again. Seems weird for a `find` method to raise `ValueError`.

Fokko replied:
See other comment

rdblue commented on python/pyiceberg/table/__init__.py:
```diff
@@ -934,206 +963,611 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
+        new_id = self.assign_new_column_id()
+
+        # update tracking for moves
+        self._added_name_to_id[full_name] = new_id
```
Don't we also need to update `_id_to_parent`?

Fokko replied:
Good one, added! The `_id_to_parent` visitor is part of https://github.com/apache/iceberg/pull/8393

rdblue commented on python/pyiceberg/table/__init__.py:
```diff
@@ -934,206 +963,611 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
+
+        return self._added_name_to_id.get(name)
+
+    def _move(self, full_name: str, move: Move) -> None:
```
Why not keep `full_name` in `Move`?

Fokko replied:
I just copied it from Java without any modifications, but I like it!

rdblue commented on python/pyiceberg/table/__init__.py:
```diff
@@ -934,206 +963,611 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
+            The UpdateSchema with the move operation staged.
+        """
+        path = (path,) if isinstance(path, str) else path
+        full_name = ".".join(path)
```
Could dedup by moving this into `_find_for_move`.

Fokko replied:
It looks like we also need this in the exceptions, but I don't think we need the paths. Cleaned this up a bit.

rdblue commented on python/pyiceberg/table/__init__.py:
```diff
@@ -934,206 +963,611 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
+
+        return self
+
+    def move_after(self, path: Union[str, Tuple[str, ...]], after_name: Union[str, Tuple[str, ...]]) -> UpdateSchema:
```
This could be smaller if you used an `_internal_move(self, path, reference, op)` method.

Fokko replied:
I played around with this, but it gets messy because it also requires the `reference`, which also uses `_find_for_move`. I cleaned it up abit.

rdblue commented on python/pyiceberg/table/__init__.py:
```diff
@@ -934,206 +963,611 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
+
+        return self
+
+    def allow_incompatible_changes(self) -> UpdateSchema:
```
I think it is probably cleaner not to set this using an API method. If it is only set in the constructor then it must be constant for the entire update operation.

Fokko replied:
I agree, should we do the same for case sensitivity?

rdblue commented on python/pyiceberg/table/__init__.py:
```diff
@@ -934,206 +963,611 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
+        new_schema = self._apply()
+
+        if new_schema != self._schema:
+            last_column_id = max(self._schema.highest_field_id, new_schema.highest_field_id)
```
This should also be the table's last_assigned_field_id.

rdblue commented on python/pyiceberg/table/__init__.py:
```diff
@@ -934,206 +963,611 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
+    ) -> Optional[IcebergType]:
         key_id: int = map_type.key_field.field_id
-        if key_id in self.adds:
+        if key_id in self._adds:
```
I think there are a lot more checks here in the Java version.

Fokko replied:
Yes, added these as well

rdblue commented on python/pyiceberg/table/__init__.py:
```diff
@@ -934,206 +963,611 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
-        return primitive
+def _add_fields(fields: Tuple[NestedField, ...], adds: Optional[List[NestedField]]) -> Optional[Tuple[NestedField, ...]]:
+    adds = adds or []
+    return None if len(adds) == 0 else fields + tuple(adds)
```
The way this is called, do we want to return None? `_add_and_move_fields` doesn't check for it.

Fokko replied:
No, I think we can drop the `None`

Fokko commented on python/pyiceberg/table/__init__.py:
```diff
@@ -889,28 +888,58 @@ def to_ray(self) -> ray.data.dataset.Dataset:
-        self._last_column_id = itertools.count(schema.highest_field_id + 1)
-        self._identifier_field_names = schema.column_names
+        self._schema = table.schema()
+        self._last_column_id = itertools.count(self._schema.highest_field_id + 1)
```
Thanks, I also ran into this in a test :)

Fokko commented on python/pyiceberg/table/__init__.py:
```diff
@@ -934,206 +963,611 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
+        Returns:
+            The UpdateSchema with the rename operation staged.
+        """
+        name_from = tuple(path_from.split(".")) if isinstance(path_from, str) else path_from
```
This was for updating the identifier fields. I've updated that to:
```python
        if path_from in self._identifier_field_names:
            self._identifier_field_names.remove(path_from)
            self._identifier_field_names.add(f"{path_from[:-len(field_from.name)]}{new_name}")
```

rdblue commented on python/mkdocs/docs/api.md:
```diff
@@ -133,52 +118,140 @@ partition_spec = PartitionSpec(
+static_table = StaticTable.from_metadata(
+    # For example:
+    # "s3a://warehouse/wh/nyc.db/taxis/metadata/00002-6ea51ce3-62aa-4197-9cf8-43d07c3440ca.metadata.json",
+    tbl.metadata_location,
```
What's `tbl` here?

Fokko replied:
This is the location of a table that we created earlier.

Some context: I was also trying to run tests on the docs, so we know that they are up to date. I will split that out to a different PR since that was much more work than anticipated, and this PR has grown quite big already. The `s3a://` location in the comment above will change on each run, so just plugging in `tbl.metadata_location` created in an earlier example would fix this.

rdblue commented on python/mkdocs/docs/api.md:
```diff
@@ -133,52 +118,140 @@ partition_spec = PartitionSpec(
+    # "s3a://warehouse/wh/nyc.db/taxis/metadata/00002-6ea51ce3-62aa-4197-9cf8-43d07c3440ca.metadata.json",
+    tbl.metadata_location,
+    properties={
+        "s3.endpoint": "http://127.0.0.1:9000",
```
Probably not needed

Fokko replied:
This was for running the docs as a test, this would then connect to the minio instance :)

rdblue commented on python/mkdocs/docs/api.md:
```diff
@@ -133,52 +118,140 @@ partition_spec = PartitionSpec(
+
+static_table = StaticTable.from_metadata(
+    # For example:
+    # "s3a://warehouse/wh/nyc.db/taxis/metadata/00002-6ea51ce3-62aa-4197-9cf8-43d07c3440ca.metadata.json",
```
`s3a`?

Fokko replied:
Historical reasons: https://github.com/tabular-io/docker-spark-iceberg/pull/91

rdblue commented on python/mkdocs/docs/api.md:
```diff
@@ -133,52 +118,140 @@ partition_spec = PartitionSpec(
+with table.update_schema() as update:
+    update.move_first("symbol")
+    update.move_after("bid", "ask")
+    # In a struct, only the new name field
```
What do you mean here?

Fokko replied:
That you don't have to provide the full struct path in the `to` argument, I've rephrased it.

rdblue commented on python/pyiceberg/table/__init__.py:
```diff
@@ -934,206 +976,630 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
+                parent_field = parent_type.element_field
+
+            if not parent_field.field_type.is_struct:
+                raise ValueError(f"Cannot add column '{name}' to non-struct type: {'.'.join(parent)}")
```
Nit: the second interpolation could be replaced with `{parent_full_path}`.

Fokko replied:
I like that, thanks

rdblue commented on python/pyiceberg/table/__init__.py:
```diff
@@ -934,206 +976,630 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
-        """Apply the pending changes to the original schema and returns the result.
+        if path_from in self._identifier_field_names:
+            self._identifier_field_names.remove(path_from)
+            self._identifier_field_names.add(f"{path_from[:-len(field_from.name)]}{new_name}")
```
I thought this was suspicious at first, but it's actually correct. Nice fix! We'll have to add the same to Java.

rdblue commented on python/pyiceberg/table/__init__.py:
```diff
@@ -934,206 +976,630 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
+
+        return self
+
+    def require_column(self, path: Union[str, Tuple[str, ...]]) -> UpdateSchema:
```
Isn't this `update_column('a', required=True)`? Is this intended to be a nicer way to set that?

rdblue replied:
Ah, I see that this is using the same internals. Looks good to me.

rdblue commented on python/pyiceberg/table/__init__.py:
```diff
@@ -934,206 +976,630 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
+        Returns:
+            The UpdateSchema with the requirement change staged.
+        """
+        self._set_column_requirement(path, False)
```
Nit: `required=False`

Fokko replied:
Good one, thanks

rdblue commented on python/pyiceberg/table/__init__.py:
```diff
@@ -934,206 +976,630 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
-        field = NestedField(new_id, name, new_type, not is_optional, doc)
+        return self
+
+    def update_column_doc(self, path: Union[str, Tuple[str, ...]], doc: str) -> UpdateSchema:
```
It's fine to have this and the methods to set the requirement for a field, but it doesn't seem very pythonic to have `update_column_doc` and `require_column`, as well as `update_column` because it creates more than one way to do the same task. Plus it's a bit more code. Definitely up to you whether you think it's a good idea to keep these.

Fokko replied:
I'll clean them up, less is more

rdblue commented on python/pyiceberg/table/__init__.py:
```diff
@@ -934,206 +976,630 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
+                except ResolveError as e:
+                    raise ValidationError(f"Cannot change column type: {full_name}: {field.field_type} -> {field_type}") from e
+
+        if updated := self._updates.get(field.field_id):
```
Since there are other methods to add a new change or update the pending change for `required` and `doc`, I think this should go inside the `field_type is not None` branch above. That way, each change is independently checked to see if it is necessary and no change is created if it is not.

The way it currently works, there will always be an update created, even if `field_type` does not change because this is outside of the checks.

Fokko replied:
I agree, I removed the other fields

Fokko replied:
Also added a check to make the noop possible, and added a test

rdblue commented on python/pyiceberg/table/__init__.py:
```diff
@@ -934,206 +976,630 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
+                    raise ValueError(f"Cannot move field {move.full_name} to a different struct")
+
+            self._moves[parent_field.field_id] = self._moves.get(parent_field.field_id, []) + [move]
+        else:
```
top-level field right?

Fokko replied:
I'm missing the point here :(

Fokko replied:
This move is on the top level indeed, added a comment

rdblue commented on python/pyiceberg/table/__init__.py:
```diff
@@ -934,206 +976,630 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
+                if self._id_to_parent.get(move.field_id) != self._id_to_parent.get(move.other_field_id):
+                    raise ValueError(f"Cannot move field {move.full_name} to a different struct")
+
+            self._moves[parent_field.field_id] = self._moves.get(parent_field.field_id, []) + [move]
```
Above in `add_column` you use an `if` for a very similar case, but I'm good either way.

rdblue commented on python/pyiceberg/table/__init__.py:
```diff
@@ -934,206 +976,630 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
+        if field_id is None:
+            raise ValueError(f"Cannot move missing column: {full_name}")
+
+        after_full_name = ".".join(after_name) if isinstance(after_name, tuple) else after_name
```
Nit: in `move_before` the arg name is `before_path` but it is `after_name` here. Probably should use a consistent name: `after_path`.

Fokko replied:
I agree, updated it! 👍🏻

rdblue commented on python/pyiceberg/table/__init__.py:
```diff
@@ -934,206 +976,630 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
+
+            if self._transaction is not None:
+                self._transaction._append_updates(*updates)  # pylint: disable=W0212
+                self._transaction._append_requirements(*requirements)  # pylint: disable=W0212
```
Not a problem, but note that if you run multiple schema updates in the same transaction, the transaction needs to ensure it sends only one `AssertCurrentSchemaId`. Right now that will fail but in the future we'll want to suppress the additional ones so that only the first schema ID is validated.

Fokko replied:
Great observation. I checked and we only allow a single `AddSchemaUpdate` in a transaction. So I think we're okay for now. I've added a test to capture the behavior.

rdblue commented on python/pyiceberg/table/__init__.py:
```diff
@@ -934,206 +976,630 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
+            raise ValueError(f"Cannot update a column that will be deleted: {full_name}")
+
+        if field_type is not None:
+            if not self._allow_incompatible_changes and field.field_type != field_type:
```
The Java equivalent ensures that only primitive fields are updated because the only type that can be passed in is a `PrimitiveType`. I think we need to have an additional check here that this is not replacing a nested type.

It may be that `promote` already does this for us, but I want to highlight that we cannot allow calling this to update a `struct`, `map`, or `list` type.

Fokko replied:
Schema evolution catches this:
```
Cannot change column type: foo: struct<2: bar: optional string> -> string
```
I've added a check to make the error more specific error:
```
Cannot change column type: struct<2: bar: optional string> is not a primitive
```

rdblue commented on python/pyiceberg/table/__init__.py:
```diff
@@ -934,206 +976,630 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
+def _add_and_move_fields(
+    fields: Tuple[NestedField, ...], adds: List[NestedField], moves: List[Move]
+) -> Optional[Tuple[NestedField, ...]]:
+    if adds:
```
Do the calls to this method need `adds or []` and `moves or []` since this check is done?

Fokko replied:
This will check the length. I prefer to pass in empty lists instead of None's, but I think we need to update `adds` to `len(adds) > 0` to make it more explicit.

rdblue commented on python/tests/test_integration_schema.py:
```diff
@@ -0,0 +1,2617 @@
+
+
+@pytest.fixture()
+def catalog() -> Catalog:
```
Why are all of these tests done in integration tests? In Java, we test the schema update API by calling `apply` to produce the schema that will be committed. Only the commit needs to be done in an integration test that way.

Fokko replied:
For me it adds another layer of testing since the REST catalog will also reject certain things (for example, the requirements on the identifier fields), this caught already a few bugs. I also like to follow the same path that the end user would take

rdblue replied:
Okay, I get it now. Seems alright, although the REST catalog really doesn't check much. It would probably be good to move most of the schema evolution behavior tests to unit tests, but the important thing is that they're happening somewhere.

rdblue commented on python/tests/test_integration_schema.py:
```diff
@@ -0,0 +1,2617 @@
+def test_add_column(simple_table: Table) -> None:
+    update = UpdateSchema(simple_table)
+    update.add_column(path="b", field_type=IntegerType())
+    apply_schema: Schema = update._apply()  # pylint: disable=W0212
```
Looks like this may not need to be done in an integration test.

rdblue commented on python/tests/test_integration_schema.py:
```diff
@@ -0,0 +1,2617 @@
+
+
+@pytest.mark.integration
+@pytest.mark.integration
```
duplication?

Fokko replied:
Added those using search-replace 🙈

rdblue commented on python/tests/test_integration_schema.py:
```diff
@@ -0,0 +1,2617 @@
+        NestedField(field_id=3, name="col_string", field_type=StringType(), required=False),
+        NestedField(field_id=4, name="col_integer", field_type=IntegerType(), required=False),
+        NestedField(field_id=5, name="col_long", field_type=LongType(), required=False),
+        schema_id=1,
```
Each schema change should produce a new schema ID in table metadata. If you were testing just `_apply` then this would be correct, but since this is actually committed, the schema ID needs to be assigned for every change.

@Fokko, this is a blocker.

Fokko replied:
The schema-IDs are updated correctly, but we don't validate them. The Python `__eq__` is the same as Java's schema: https://github.com/apache/iceberg/blob/master/api/src/main/java/org/apache/iceberg/Schema.java#L464-L473

Fokko replied:
![Uploading image.png…]()

rdblue commented on python/tests/test_integration_schema.py:
```diff
@@ -0,0 +1,2617 @@
+    FixedType(3),
+    FixedType(4),
+    # We'll just allow Decimal promotions right now
+    # https://github.com/apache/iceberg/issues/8389
```
Changing the scale of a decimal is not an allowed promotion.

Fokko replied:
Yes it passes right now :)

rdblue commented on python/tests/test_integration_schema.py:
```diff
@@ -0,0 +1,2617 @@
+
+
+@pytest.mark.integration
+def test_rename_simple_nested(catalog: Catalog) -> None:
```
Would be nice to have a case with `.` names, like `rename_column(("a.b", "c.d"), "e.f")` produces column `("a.b", "e.f")`.

Fokko replied:
Added

rdblue commented on python/tests/test_integration_schema.py:
```diff
@@ -0,0 +1,2617 @@
+            ),
+            NestedField(field_id=4, name="foo", field_type=StringType(), required=True),
+            schema_id=0,
+            identifier_field_ids=[],
```
Could add identifier fields to test nested identifier rename here.

Fokko replied:
This was a great test case. With case-insensitive the identifier-field-ids were not properly updated

rdblue reviewed (changes requested):
Looks really good to me! I think the only issues are:

rdblue commented:
Awesome work, @Fokko! This is really close.

Fokko commented:
> Schema ID is not being assigned in commit, which corrupts metadata

This was a great catch, updated this! The REST Catalog actually handled this correctly, but now it is also fixed and tested outside of the rest catalog.

> Tests are run as integration tests, but most could be converted to unit tests that call _apply

I ported some tests outside of the integration tests as well. As [mentioned above](https://github.com/apache/iceberg/pull/8374#discussion_r1314249767) I really like testing against the REST catalog, since uses the same public API that the customer uses (which is a good thing from an OOP perspective), but we're testing much more code than we actually need. I also ported some test back to avoid having only integration tests, but the majority is still IT tests because I prefer them. Let me know if you want me to port the rest back as well.

> update_column always creates an entry in changes even if the change should be a noop

Added and tested as well, also removed the `make_required` and `set_doc` operations.

Fokko commented:
I accidentally tagged this PR in another PR that was closed.

rdblue reviewed (approved):
Looks good to me. I don't understand why so many schema tests are run as integration tests, but at least they run somewhere. We can get this in to unblock the release at least.
